verification/src/solidity/github_fetcher.rs

```diff
@@ -0,0 +1,191 @@
+use crate::compiler::{
+    fetcher::Fetcher,
+    version::{CompilerVersion, ParseError},
+};
+use async_trait::async_trait;
+use std::{
+    collections::HashMap,
+    fs::{File, OpenOptions},
+    io::ErrorKind,
+    os::unix::prelude::OpenOptionsExt,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+use thiserror::Error;
+use url::Url;
+
+#[derive(Default)]
+pub struct GithubFetcher {
+    releases: HashMap<CompilerVersion, url::Url>,
+    folder: PathBuf,
+}
+
+#[derive(Error, Debug)]
+pub enum ListError {
+    #[error("fetching github returned error: {0}")]
+    Github(octocrab::Error),
+    #[error("error parsing tag name into version: {0}")]
+    TagName(ParseError),
+}
+
+impl GithubFetcher {
+    async fn list_releases(
+        owner: &str,
+        repo: &str,
+    ) -> Result<HashMap<CompilerVersion, Url>, ListError> {
+        let octocrab = octocrab::instance();
+        let mut page_index = 0u32;
+        let mut releases = HashMap::default();
+
+        // we can make this faster by fetching the pages in parallel
+        loop {
+            let page = octocrab
+                .repos(owner, repo)
+                .releases()
+                .list()
+                .per_page(100)
+                .page(page_index)
+                .send()
+                .await
+                .map_err(ListError::Github)?;
+
+            if page.items.is_empty() {
+                break;
+            }
+
+            for release in page {
+                let solc = release
+                    .assets
+                    .into_iter()
+                    .find(|asset| asset.name == "solc");
+                if let Some(solc) = solc {
+                    let version =
+                        CompilerVersion::from_str(&release.tag_name).map_err(ListError::TagName)?;
+                    releases.insert(version, solc.browser_download_url);
+                }
+            }
+
+            page_index += 1;
+        }
+
+        Ok(releases)
+    }
```

Comment:
Cannot we do the same way it is in svm - maintain _list.json_ (https://solc-bin.ethereum.org/linux-amd64/list.json) in the repository which is updated every time new release is added, and download and parse only that?

Reply:
why?
you see, if we scan the releases, we get 100% guarantee, that they are valid (yes, you could scan and then delete the release but I'm talking about the other stuff)

while we need to constantly update the file: we need some way to add new releases to it in cicd, if for some reason we delete a release (may be it's broken or something), we need to manually change the file, while doing it we may forget to add/remove the comma in json or something and now the file is invalid

imo that's just a lot of extra work for nothing (actually for a lot of places where we can make an error)

why they do it? if you check the file carefully, you'll see that it actually have another links to the external file storage, which we don't and probably won't have (if we will, let's talk about it then) 

also svm doesn't use this file, it uses some data that actually comes with the svm cargo crate + constructs the url magically (because it downloads from repo, not the releases), which I think is actually wrong way to do it

Reply:
In my opinion, that may have 3 main advantages:
1. Iterating all releases from the repository may be quite a long process. As we have to do that on each startup that may considerable increase the cost of running a new instance. Additionally, if _github_ will return non-success response on some of our requests, it may create an inconsistency between versions our client considers as available and versions actually available (and, thus, with other instances).At the same time, downloading only one file on startup is cheap and we may just retry or not to run an instance at all in case of any problems.
2. We are binding ourselves to _github_ and its releases mechanism in that case. When specifying urls to download compilers in a separate file, we are free to keep them wherever we want (e.g. we may freely move to an S3 instance or another SVC provider). We may even keep part of the compilers at a different place from others (yet not sure whether it may be useful).
3. It is not quite clear how we would update the list of available compilers when the node is running. While when using a separate file we most probably will just have a periodic job which redownloads and parses a `list.json` file.

Reply:
I see what you are talking about, but it seems that we should decide between full consistency (when we forget to update `list.json`) and swiftness (when we have to iterate through all releases and github pages on each startup), and, in my opinion, it is better to sacrifice consistency here, as it is something we may ensure by other means.

Reply:
Though probably we may discuss that in a different issue